### PR TITLE
[#4] As a user, I can use the application to make a network request

### PR DESCRIPTION
**Pull request by luongvo**
https://github.com/nimblehq/flutter_templates/issues/4

## What happened 👀

The network request is a must to every application nowadays, so we should prepare this for our own convenience later.

## Insight 📝

- Setup Dio, Retrofit
- Setup different Base API Endpoint depending on each different flavor: `REST_API_ENDPOINT` in .env
- Implement a sample API request with Retrofit: https://jsonplaceholder.typicode.com/users, same to the sample request in our Android [CoroutineTemplate](https://github.com/nimblehq/android-templates/blob/develop/CoroutineTemplate/data/src/main/java/co/nimblehq/coroutine/data/service/ApiService.kt#L8-L9).

## Proof Of Work 📹

Sample API request and response:

```
I/flutter (11443): *** Request ***
I/flutter (11443): uri: https://jsonplaceholder.typicode.com/users
I/flutter (11443): method: GET
I/flutter (11443): responseType: ResponseType.json
I/flutter (11443): followRedirects: true
I/flutter (11443): connectTimeout: 3000
I/flutter (11443): sendTimeout: 0
I/flutter (11443): receiveTimeout: 5000
I/flutter (11443): receiveDataWhenStatusError: true
I/flutter (11443): extra: {}
I/flutter (11443): headers:
I/flutter (11443): data:
I/flutter (11443): {}
I/flutter (11443): 
I/flutter (11443): *** Response ***
I/flutter (11443): uri: https://jsonplaceholder.typicode.com/users
I/flutter (11443): statusCode: 200
I/flutter (11443): headers:
I/flutter (11443):  x-ratelimit-reset: 1633909569
I/flutter (11443):  date: Sun, 05 Dec 2021 16:19:54 GMT
I/flutter (11443):  transfer-encoding: chunked
I/flutter (11443):  x-ratelimit-limit: 1000
I/flutter (11443):  vary: Origin, Accept-Encoding
I/flutter (11443):  content-encoding: gzip
I/flutter (11443):  x-ratelimit-remaining: 999
I/flutter (11443):  pragma: no-cache
I/flutter (11443):  server: cloudflare
I/flutter (11443):  cf-ray: 6b8e99482e866e5e-HKG
I/flutter (11443):  etag: W/"160d-1eMSsxeJRfnVLRBmYJSbCiJZ1qQ"
I/flutter (11443):  connection: keep-alive
I/flutter (11443):  cache-control: max-age=43200
I/flutter (11443):  age: 13471
I/flutter (11443):  cf-cache-status: HIT
I/flutter (11443):  report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=DnmWry3OHLBZMCwBhU12ZitDEdePGN%2FPqBH8ZUZxb6Vy0zQs0mo2m3QoHUcmXDmU3SrVMhc6fdi3yiBQncQWY8OVWeSSFT1gpvAvfg%2Bb2HG7MW2rqqUxrWLPoppz%2BQwEqh57ie94EUZwuVRrR5qI"}],"group":"cf-nel","max_age":604800}
I/flutter (11443):  content-type: application/json; charset=utf-8
I/flutter (11443):  expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
I/flutter (11443):  access-control-allow-credentials: true
I/flutter (11443):  x-powered-by: Express
I/flutter (11443):  alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400, h3-28=":443"; ma=86400, h3-27=":443"; ma=86400
I/flutter (11443):  nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
I/flutter (11443):  via: 1.1 vegur
I/flutter (11443):  x-content-type-options: nosniff
I/flutter (11443):  expires: -1
I/flutter (11443): Response Text:
I/flutter (11443): [{id: 1, name: Leanne Graham, username: Bret, email: Sincere@april.biz, address: {street: Kulas Light, suite: Apt. 556, city: Gwenborough, zipcode: 92998-3874, geo: {lat: -37.3159, lng: 81.1496}}, phone: 1-770-736-8031 x56442, website: hildegard.org, company: {name: Romaguera-Crona, catchPhrase: Multi-layered client-server neural-net, bs: harness real-time e-markets}}, {id: 2, name: Ervin Howell, username: Antonette, email: Shanna@melissa.tv, address: {street: Victor Plains, suite: Suite 879, city: Wisokyburgh, zipcode: 90566-7771, geo: {lat: -43.9509, lng: -34.4618}}, phone: 010-692-6593 x09125, website: anastasia.net, company: {name: Deckow-Crist, catchPhrase: Proactive didactic contingency, bs: synergize scalable supply-chains}}, {id: 3, name: Clementine Bauch, username: Samantha, email: Nathan@yesenia.net, address: {street: Douglas Extension, suite: Suite 847, city: McKenziehaven, zipcode: 59590-4157, geo: {lat: -68.6102, lng: -47.0653}}, phone: 1-463-123-4447, website: ramiro.info, company: {name: Romag
```

### CODE REVIEW

File: pubspec.yaml [hunk 1]
```diff
@@ -22,22 +22,29 @@ environment:
 
 dependencies:
   cupertino_icons: ^1.0.2
+  dio: ^4.0.0
   flutter:
     sdk: flutter
   flutter_localizations:
     sdk: flutter
   flutter_config: ^2.0.0
+  freezed: ^0.14.1+3
```
Comment (manh-t) [inline pubspec.yaml:31]: @luongvo please move `freezed` to `dev_dependencies` section by following [this tutorial](https://pub.dev/packages/freezed#how-to-use) 😄

File: pubspec.yaml [hunk 2]
```diff
@@ -22,22 +22,29 @@ environment:
...
+  freezed_annotation: ^0.14.1
+  injectable: ^1.3.0
```
Comment (manh-t) [inline pubspec.yaml:33]: it seems we are missing `get_it` and `injectable_generator` packages. [Reference](https://pub.dev/packages/injectable#installation) 😅

Comment (luongvo) [reply pubspec.yaml:33]: @manh-t we don't implement the DI part in this task so adding get_it and injectable_generator are unnecessary

File: pubspec.yaml (hunk 1 above)
Comment (luongvo) [reply pubspec.yaml]: Updated

File: pubspec.yaml (hunk 2 above)
Comment (manh-t) [reply pubspec.yaml]: @luongvo shall we remove `network_module`? It's getting error there 💅

Comment (luongvo) [reply pubspec.yaml]: @manh-t agree to remove

**Review by manh-t — APPROVED** [review]
LGTM 🚀

**Review by sleepylee — APPROVED** [review]

**Pull request merged**